# The main entry point of your workflow.
# After configuring, running snakemake -n in a clone of this repository should successfully execute a dry-run of the workflow.


report: "report/workflow.rst"

# Allow users to fix the underlying OS via singularity.
singularity: "docker://continuumio/miniconda3"


        # The first rule should define the default target files
        # Subsequent target rules can be specified below. They should start with all_*.
# barcodes = glob_wildcards("results/{barcode}/medaka/consensus.fasta ")
# rule all:
#     input:
#         expand("results/{barcode}/medaka/consensus.fasta", barcode=barcodes)

include: "rules/common.smk"


rule porechop_adapter_barcode_trimming:
    input:
        fastq_in="input_data/{barcode}.fastq"
    output:
        out_dir=directory("results/{barcode}"),
        out_file="results/{barcode}/{barcode}.abtrim.fastq"
        #"results/{sample}/{sample}_pc.fastq"
    conda:
        "envs/porechop.yaml"
    threads: 8
    shell:
        "porechop --require_two_barcodes -i {input.fastq_in} -b {output.out_dir} --threads {threads}"


rule convert2fasta:
    input:
        "results/{barcode}/{barcode}.abtrim.fastq"
    output:
        "results/{barcode}/{barcode}.abtrim.fasta"
    conda:
        "envs/seqtk.yaml"
    shell:
        "seqtk seq -A  {input} > {output}"


rule canu_correct:
    input:
        "results/{barcode}/{barcode}.abtrim.fasta"
    output:
        out_dir=directory("results/{barcode}"),
        out_file="results/{barcode}/{barcode}.correctedReads.fasta.gz"
    conda:
        "envs/canu.yaml"
    threads: 8
    shell:
        "canu -correct -nanopore-raw {input} -p {wildcards.barcode} -d {output.outdir} genomeSize=30k corOverlapper=minimap utgOverlapper=minimap obtOverlapper=minimap minOverlapLength=10 minReadLength=200 corMMapMerSize=10 corOutCoverage=50000 corMinCoverage=0 maxInputCoverage=20000"


rule medaka_polishing:
    input:
        "results/{barcode}/{barcode}.fastq"
    output:
        outdir=directory("results/{barcode}/medaka"),
        outfile="results/{barcode}/medaka/consensus.fasta"
    conda:
        "envs/medaka.yaml"
    threads: 2
    shell:
        "medaka_consensus -i {input} -o {output.outdir} -d resources/MN908947.3_SARS-CoV2_Wuhan-reference.fasta -t {threads}"