# The main entry point of your workflow.
# After configuring, running snakemake -n in a clone of this repository should successfully execute a dry-run of the workflow.
import os

report: "report/workflow.rst"

# Allow users to fix the underlying OS via singularity.
singularity: "docker://continuumio/miniconda3"


        # The first rule should define the default target files
        # Subsequent target rules can be specified below. They should start with all_*.
#barcodes = glob_wildcards("results/{barcode}/medaka/consensus.fasta ")
in_files = [fq_file.path for fq_file in os.scandir("./input_data") if os.path.isfile(fq_file)]
accept_ext = ["fastq", "fq", "fasta", "fa", "fna"]
barcodes = [os.path.split(fpath)[-1].split(".")[0] for fpath in in_files if os.path.split(fpath)[-1].split(".")[1] in accept_ext]
print(barcodes)
rule all:
    input:
        #expand("results/{barcode}_corr/{barcode}.correctedReads.fasta.gz", barcode=barcodes),
        expand("results/{barcode}_polish/medaka/consensus.fasta", barcode=barcodes),
        expand("results/{barcode}_primertrim/{barcode}.fastq", barcode=barcodes)
        #input3="results/report.txt"



#include: "rules/common.smk"
#include: "rules/get_samples.smk"

rule intial_qc:
    input:
        "input_data/{barcode}.fastq"
    output:
        directory("results/{barcode}/qc/initial")
    log:
        "logs/{barcode}/{barcode}_initialQC.log"
    conda:
        "envs/nanoqc.yaml"
    shell:
        "nanoQC {input} -o {output}"


rule porechop_adapter_barcode_trimming:
    input:
        fastq_in="input_data/{barcode}.fastq"
    output:
        out_dir=directory("results/{barcode}_abtrim"),
        out_file="results/{barcode}_abtrim/{barcode}.fastq"
        #"results/{sample}/{sample}_pc.fastq"
    conda:
        "envs/porechop.yaml"
    threads: 2
    shell:
        "porechop -i {input.fastq_in} -o {output.out_file} --threads {threads} -v 0"


rule customize_primer_porechop:
    output:
        "workflow/report/replacement_notice.txt"
    conda:
        "envs/primechop.yaml"
    shell:
        """
        rm $CONDA_PREFIX/lib/python3.6/site-packages/porechop/adapters.py
        mv ./resources/ARTIC_v3_adapters.py $CONDA_PREFIX/lib/python3.6/site-packages/porechop/adapters.py
        touch workflow/report/replacement_notice.txt
        echo "replaced adpaters in adapter.py file with ARTICv3 primers" > workflow/report/replacement_notice.txt
        """


rule porechop_primer_trimming:
    input:
        fastq_in="results/{barcode}_abtrim/{barcode}.fastq",
        repl_flag="workflow/report/replacement_notice.txt"
    output:
        "results/{barcode}_primertrim/{barcode}.fastq"
    conda:
        "envs/primechop.yaml"
    threads: 2
    shell:
        "porechop -i {input.fastq_in} -o {output} --no_split --end_size 35 --extra_end_trim 0 --threads {threads} -v 0"


rule nanofilt:
    input:
        "results/{barcode}_primertrim/{barcode}.fastq"
    output:
        "results/{barcode}/trim-filt/{barcode}_tf.fastq"
    log:
        "logs/{barcode}/{barcode}_nanofilt.log"
    conda:
        "envs/nanofilt.yaml"
    shell:
        "NanoFilt {input} -q 10 --maxlength 500 > {output}"


rule post_trimfilt_qc:
    input:
        "results/{barcode}/trim-filt/{barcode}_tf.fastq"
    output:
        directory("results/{barcode}/qc/post_trimfilt_qc")
    log:
        "logs/{barcode}/{barcode}_post_trimfilt_qc.log"
    conda:
        "envs/nanoqc.yaml"
    shell:
        "nanoQC {input} -o {output}"


rule convert2fasta:
    input:
        #"results/{barcode}_abtrim/{barcode}.fastq"
        "results/{barcode}/trim-filt/{barcode}_tf.fastq"
    output:
        "results/{barcode}_fa/{barcode}_tf.fasta"
    conda:
        "envs/seqtk.yaml"
    shell:
        "seqtk seq -A  {input} > {output}"


rule medaka_polishing:
    input:
        in_file="results/{barcode}_fa/{barcode}_tf.fasta"
        #in_file="results/{barcode}_abtrim/{barcode}.fastq",
        #convcompl="results/{barcode}_ab_fa/{barcode}.fasta"
    output:
        outdir=directory("results/{barcode}_polish/medaka"),
        outfile="results/{barcode}_polish/medaka/consensus.fasta"
    conda:
        "envs/medaka.yaml"
    threads: 4
    shell:
        "medaka_consensus -i {input.in_file} -o {output.outdir} -d resources/MN908947.3_SARS-CoV2_Wuhan-reference.fasta -t {threads}"


rule canu_correct:
    input:
        "results/{barcode}_fa/{barcode}_tf.fasta"
    output:
        out_dir=directory("results/{barcode}_corr"),
        out_file="results/{barcode}_corr/{barcode}.correctedReads.fasta.gz"
    conda:
        "envs/canu.yaml"
    threads: 6
    shell:
        #"canu -correct -nanopore-raw {input} -p {wildcards.barcode} -d {output.out_dir} genomeSize=30k corOverlapper=minimap utgOverlapper=minimap obtOverlapper=minimap minOverlapLength=10 minReadLength=200 corMMapMerSize=10 corOutCoverage=50000 corMinCoverage=0 maxInputCoverage=20000"
        "canu -correct -nanopore {input} -p {wildcards.barcode} -d {output.out_dir} genomeSize=30k corOverlapper=minimap utgOverlapper=minimap obtOverlapper=minimap minOverlapLength=10 minReadLength=200 corMMapMerSize=10 corOutCoverage=50000 corMinCoverage=0 maxInputCoverage=20000"


# rule write_report:
#     input:
#         # input1=expand("results/{barcode}/medaka/consensus.fasta", barcode=barcodes),
#         # input2=expand("results/{barcode}/{barcode}.correctedReads.fasta.gz", barcode=barcodes)
#         input1="results/{barcode}/medaka/consensus.fasta",
#         input2="results/{barcode}/{barcode}.correctedReads.fasta.gz"
#     output:
#         report="results/report.txt"
#     script:
#         "./scripts/write_report.py {output.report}{input1},{input2}"
        